### PR TITLE
[CuTe,Flex] varlen blocksparsity

### PR DESCRIPTION
**Pull request by reubenconducts**
This PR extends blocksparsity to the variable sequence length case. Whereas in batched blocksparsity the metadata tensors take the shapes
```python
*_block_cnt: [batch_size, num_heads, num_m_blocks]
*_block_idx: [batch_size, num_heads, num_m_blocks, num_n_blocks],
```
in varlen blocksparsity, we pack our metadata tensors to take the shapes
```python
*_block_cnt: [num_heads, total_m_blocks]
*_block_idx: [num_heads, total_n_blocks]
```
where `total_m_blocks` is the sum of all `m` blocks per head (equiv. number of work tiles per head) and `total_n_blocks` is the total of all `n` blocks potentially processed per head across all sequences in the batch. For example, consider a varlen batch with sequences contained in `seqlens_q` and `seqlens_k`. At batch index `b`, we let 
```tex
num_m(b) = ceildiv(seqlens_q[b], tile_m)
num_n(b) = ceildiv(seqlens_k[b], tile_n)
```
and define
```tex
total_m_blocks = sum_{b \in B} num_m(b)
total_n_blocks = sum_{b \in B} num_m(b) * num_n(b)
```

To properly index into the blocksparsity tensors, we use auxiliary `mCuTotalMBlocks` and `mBlockIdxOffsets` tensors, which can be prepared on host. 

cc @drisspg @v0i0 

NOT INTENDED FOR THIS PR:
- bwd support

### CODE REVIEW

**Comment by SeanLi-OI** [comment]
Hi there, @reubenconducts ! Thank you so much for your draft.

Since I also need this feature eagerly, I tried to continue development based on your branch which fixes some grammar issue (https://github.com/SeanLi-OI/flash-attention/commit/7ccfc5eeef27cbc4aadcc44bbea4131f0549f8af). Though it can run, but returns wrong results when batch_size > 1.

I completely understand you may be busy with other priorities. If you have a moment, I’d be truly grateful for any guidance:
Are my modifications heading in the right direction? Or do you plan to continue updating this PR?

**Comment by reubenconducts** [comment]
@SeanLi-OI Yes, I will be continuing this, but not until next week.

File: flash_attn/cute/block_sparse_utils.py [hunk 1]
```diff
@@ -574,6 +622,7 @@ def produce_block_sparse_loads_sm100(
     batch_idx,
     head_idx,
     m_block,
+    seqlen,
```
Comment (drisspg) [inline flash_attn/cute/block_sparse_utils.py]: Nit: why diff name here

File: flash_attn/cute/block_sparse_utils.py [hunk 2]
```diff
@@ -891,7 +950,7 @@ def softmax_block_sparse_sm100(
                     full_n_block,
                     is_first=False,
                     mask_fn=partial(
-                        mask_fn_none, mask_seqlen=False, check_q_boundary=check_m_boundary
+                        mask_fn_none, mask_seqlen=True, check_q_boundary=check_m_boundary
```
Comment (drisspg) [inline flash_attn/cute/block_sparse_utils.py:935]: Hmmm

Comment (reubenconducts) [reply flash_attn/cute/block_sparse_utils.py:935]: This is a legitimate bug, though I didn't handle it particularly well. We should *always* `mask_seqlen` the first full block, because the first masked block may not be the rightmost block processed. Somehow our tests didn't expose this properly.

File: flash_attn/cute/block_sparse_utils.py (hunk 1 above)
Comment (reubenconducts) [reply flash_attn/cute/block_sparse_utils.py]: No reason; will change to match

File: flash_attn/cute/block_sparse_utils.py (hunk 2 above)
Comment (reubenconducts) [reply flash_attn/cute/block_sparse_utils.py:935]: Actually, this is handled properly. The first (rightmost) full block is always masked.

**Comment by wqwqazwsxedc** [comment]
Hi @drisspg, @reubenconducts, just checking in on this PR. Are there any remaining blockers or changes needed before it can be merged? 
I think this feature could be useful to implement something akin to [PrefixGrouper](https://github.com/CASIA-IVA-Lab/PrefixGrouper/tree/main) for varlen sequences, which would be useful for RL training. I'd be glad to help if necessary.

File: flash_attn/cute/interface.py [hunk 3]
```diff
@@ -598,45 +601,61 @@ def _flash_attn_fwd(
...
     normalized_block_sparse_tensors = None
     q_subtile_factor = None
     if block_sparse_tensors is not None:
-        if seqlen_q is None:
-            raise ValueError("Block sparsity requires fixed-length sequences (seqlen_q must be known).")
-        (
-            normalized_block_sparse_tensors,
-            block_sparse_broadcast_pattern,
-            q_subtile_factor,
-        ) = normalize_block_sparse_config(
```
Comment (drisspg) [inline flash_attn/cute/interface.py:631]: nit: could we jam somemore of this into the normalize to keep the interface cleaner?

File: flash_attn/cute/interface.py [hunk 4]
```diff
@@ -1970,6 +2001,10 @@ def forward(
         deterministic: bool = False,
         score_mod: Optional[Callable] = None,
         score_mod_bwd: Optional[Callable] = None,
+        mask_mod: Optional[Callable] = None,
+        block_sparse_tensors: Optional[list] = None,
+        cu_total_m_blocks: Optional[torch.Tensor] = None,
```
Comment (drisspg) [inline flash_attn/cute/interface.py]: I thought this would be attributes of the BlcokSparseTensor tuple?

**Comment by drisspg** [comment]
total_m_blocks = sum_b ceil(seqlen_q[b] / tile_m)  whats the flow you envision for people producing this? Trying to minimize HtD / DtoH syncs I guess this could live in the dataloader right? Or you can overlaunch:  batch_size * ceil(max_seqlen_q / tile_m)   and then mask

File: flash_attn/cute/block_sparse_utils.py [hunk 5]
```diff
@@ -17,6 +17,48 @@
...
+
+@cute.jit
+def get_curr_blocksparse_tensors(
+    batch_idx: cutlass.Int32,
+    head_idx: cutlass.Int32,
+    m_block: cutlass.Int32,
+    blocksparse_tensors: BlockSparseTensors,
+    seqlen_info: SeqlenInfoQK,
+) -> Tuple[cutlass.Int32, cute.Tensor, cutlass.Int32, Optional[cute.Tensor]]:
+    mask_block_cnt, mask_block_idx, full_block_cnt, full_block_idx = blocksparse_tensors
```
Comment (drisspg) [inline flash_attn/cute/block_sparse_utils.py]: super nit: I think just having a varlen speicifc sub func could make this even cleaner.. I hate cutedsl huge if/else  constexpr blocks

File: flash_attn/cute/interface.py [hunk 6]
```diff
@@ -730,11 +751,13 @@ def _flash_attn_fwd(
             seqused_q_tensor,
             seqused_k_tensor,
             learnable_sink_tensor,
+            cu_total_m_blocks_tensor,
+            cu_total_n_blocks_tensor,
```
Comment (drisspg) [inline flash_attn/cute/interface.py]: Nit: cu_total_n_blocks is a bit confusing to me. The name sounds like a cumulative sum of N_b, but semantically it is a cumulative sum of M_b * N_b the offset into the packed *_block_idx tensors. Maybe something like   cu_block_idx_offsets ?

**Comment by drisspg** [comment]
Overall, sorry this took so long to review. I'm pretty on board. I did a bunch of codexing / experiments on the pr, and I think that we could build transposy-like metadata for the backward decently efficiently. I still don't have a full sense of all the different var-length sub-features this would need to interact with, but I think this provides a decent amount of value and is actually pretty well contained

We should spell out somewhere very explicitly what the mask_mod semantics are for this feature

File: flash_attn/cute/block_sparsity.py [hunk 7]
```diff
@@ -343,21 +362,38 @@ def normalize_block_sparse_config(
...
-            sparse_block_size_q=sparse_block_size_q,
-            sparse_block_size_kv=sparse_block_size_kv,
+    if tensors.cu_total_m_blocks is not None:
+        base_m_block = q_stage * m_block_size
+        if sparse_block_size_q is not None and sparse_block_size_q != base_m_block:
+            raise ValueError(
+                f"Varlen block sparsity requires sparse_block_size[0]={base_m_block} "
+                f"(= q_stage * tile_m); got {sparse_block_size_q}."
+            )
+        total_m_blocks = tensors.cu_total_m_blocks[-1].item()
```
Comment (drisspg) [inline flash_attn/cute/block_sparsity.py]: are these items expected?

Comment (drisspg) [reply flash_attn/cute/block_sparsity.py]: I dont think they are / need to be also rebase :)

**Review by drisspg — APPROVED** [review]

**Comment by drisspg** [comment]
let me know when ready to merge
